### PR TITLE
add interpolate_pos_encoding for Detr and test

### PR DESCRIPTION
**Pull request by justyx404**
# What does this PR do?

- Add interpolate position embedding to Detr model #30579
  image

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.
@amyeroberts

### CODE REVIEW

**Comment by g1y5x3** [comment]
It's a bit awkward due to Detr using sine pos embedding by default which doesn't need to be interpolated. As for as the assert in the test I made, I'm only making sure the shape stays the same as using sine pos embedding but essentially they are just the object class + bounding box prediction dim.

File: tests/models/detr/test_modeling_detr.py [hunk 1]
```diff
@@ -680,3 +680,26 @@ def test_inference_no_head(self):
             [[0.0616, -0.5146, -0.4032], [-0.7629, -0.4934, -1.7153], [-0.4768, -0.6403, -0.7826]]
         ).to(torch_device)
         self.assertTrue(torch.allclose(outputs.last_hidden_state[0, :3, :3], expected_slice, atol=1e-4))
+
+    def test_inference_interpolate_pos_encoding(self):
+        print("\n")
```
Comment (amyeroberts) [inline tests/models/detr/test_modeling_detr.py:685]:
```suggestion
```

File: tests/models/detr/test_modeling_detr.py [hunk 2]
```diff
@@ -680,3 +680,26 @@ def test_inference_no_head(self):
...
+
+        # default model uses sine position embedding
+        model = DetrModel.from_pretrained("facebook/detr-resnet-50").to(torch_device)
+        with torch.no_grad():
+            outputs_sine = model(**encoding, interpolate_pos_encoding=True)
+
+        # reassign to learned position embedding
+        model.backbone.position_embedding = DetrLearnedPositionEmbedding(model.encoder.layers[0].embed_dim // 2).to(
+            torch_device
+        )
```
Comment (amyeroberts) [inline tests/models/detr/test_modeling_detr.py:700]: Ah, yes, this makes things awkward. We can't really expect users to have to do this, nor can we assume that we can always pass `interpolate_pos_encoding` to the backbone.

**Review by amyeroberts — COMMENTED** [review]
Thanks for working on this and adding a test! 

Looking at this, I realise that this isn't something we can really add to DETR (and related models) because of the backbone in the architecture. Apologies for including it in the original list (I've just removed it from the issue), and thank you for contributing to improving our models

**Comment by g1y5x3** [comment]
totally understandable after attempting to make it work. and agree that users won't really be needing this. will close it then.
